src/switchroot/ostree-prepare-root.c

```diff
@@ -684,12 +715,60 @@ main (int argc, char *argv[])
   if (rmdir (TMP_SYSROOT) < 0)
     err (EXIT_FAILURE, "couldn't remove temporary sysroot %s", TMP_SYSROOT);
 
-  /* Now that we've set up all the mount points, if configured we remount the physical
-   * rootfs as read-only; what is visibly mutable to the OS by default is just /etc and /var.
-   * But ostree knows how to mount /boot and /sysroot read-write to perform operations.
-   */
-  if (sysroot_readonly)
+  if (sysroot_invisible)
+    {
+      /* Keep a living sysroot in a private mount namespace,
+       * and unmount sysroot in the root mount namespace to make it invisible.
+       */
+      const char *sysroot_ns = OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns";
+      glnx_autofd int ns_fd = open (sysroot_ns, O_WRONLY | O_CREAT, 0);
+      if (ns_fd < 0)
+        err (EXIT_FAILURE, "failed to create %s", sysroot_ns);
+
+      const gsize stack_size = 0x8000;
+      g_autofree void *stack = g_malloc (stack_size);
+
+      /* Block signals */
+      sigset_t oldset, newset;
+      sigfillset (&newset);
+      sigprocmask (SIG_SETMASK, &newset, &oldset);
+
+      int pid = clone (invisible_helper, (char*)stack + stack_size, CLONE_VM | CLONE_NEWNS | SIGCHLD, NULL);
+
+      sigprocmask (SIG_SETMASK, &oldset, NULL);
+
+      if (pid < 0)
+          err (EXIT_FAILURE, "failed to create child process");
+
+      /* Bind mount the private mount namespace */
```

Comment:
Huh funky...this implementation is different than I thought it'd be. I thought we'd have a "holder process" that has a file descriptor for the mount and it could either have an API to pass that fd back over a socketpair(), or callers could just `nsenter()` to its mount namespace.

(See also https://github.com/containers/bootc/pull/919 for code related to the former)

I didn't realize bind mounting a `/proc/<pid>ns/mnt` like that would work. It seems OK though to me a downside of this is that it still appears in the global filesystem state.

Hmmm...could we actually just have a stub systemd unit in the initramfs and use e.g. https://systemd.io/FILE_DESCRIPTOR_STORE/ to store the mount fd?

Reply:
> Hmmm...could we actually just have a stub systemd unit in the initramfs and use e.g. https://systemd.io/FILE_DESCRIPTOR_STORE/ to store the mount fd?

I did not know it before. Let me look into it.

Reply:
> I thought we'd have a "holder process" that has a file descriptor for the mount and it could either have an API to pass that fd back over a socketpair(), or callers could just nsenter() to its mount namespace.

A problem is to ensure that the "holder process" can live long enough. It must be alive when ostree-finalize-staged runs.

Maybe we can upload the fd to fd store and use socket activation to restart the holder when the mount namespace is required.

> It seems OK though to me a downside of this is that it still appears in the global filesystem state.

If we use `socketpair()` or `nsenter()`, every `root` process can still obtain sysroot, which does not make a difference from placing the nsfd in the global filesystem. If we want to introduce some policies, we may need dbus...

And we can use SELinux or other LSMs to prevent accessing to `/run/ostree/.private`. From a security perspective, I don't think it's really a downside.

Reply:
Anyway, I think it's too much work for ostree-prepare-root to be a holder process or communicate with systemd or dbus. If you think we need one, it should be a service outside initramfs that consumes sysroot-ns and serves it somehow.

Reply:
Let's mark this resolved and discuss this topic in main thread if you want.